### PR TITLE
Defer decompression to avoid back-pressure when cloning a response

### PR DESCRIPTION
**Pull request by Peleg**
This PR builds on top of #142 and hopefully fixes the issue where cloning a response and not reading one of the streams causes back-pressure and pauses flowing.

The issue seemed to occur only when dealing with responses which are large Transform streams (Gunzip/Inflate) which would pipe down more than they read after they transform their chunks (see: https://github.com/nodejs/node/blob/v5.5.0/lib/_stream_transform.js#L28-L37 for more info).

By deferring the decompression step until the response is being decoded, we can avoid having this issue and maintain "predictable" chunk sizes when cloning a response.

This, of course, adds the overhead of having to decompress the response each time it's being cloned.

### CODE REVIEW

**Comment by Peleg** [comment]
not sure about the test failure. they pass when run locally, even when I set the highWaterMark on the clones to be super low. is it possible it needs a longer timeout?

**Comment by bitinn** [comment]
I will have to look into it later, but I don't think the backpressure problem can be fully resolved given #151 (see linked discussion as well)

**Comment by Peleg** [comment]
Yup. I finally managed to reproduce this when cloning a (much) larger response that isnt gzipped. This PR still seems to help when the response is compressed as it pipes a much smaller stream (gzipped vs raw).

Back to the drawing board 😒

**Comment by bitinn** [comment]
Do let me know if you have come up with anything, due to lack of a good solution, we have added it to [KNOWN LIMIT document](https://github.com/bitinn/node-fetch/blob/master/LIMITS.md).

**Comment by Peleg** [comment]
thanks @bitinn. in the meantime, should there be an option to increase the PassThroughs high water mark so we dont encounter back pressure when cloning a response with body > 16KB?

**Comment by bitinn** [comment]
I don't plan to expose `highWaterMark`, here are my reasons:

- If you know you are running into this problem, the better solution is `Promise.all()`. Run your stream in parallel, don't resolve it one-after-another.
- Setting `highWaterMark` manually is not a good workaround. A good workaround is to put the stream in memory or on file. Basically, avoid `clone()` when you just need to *copy* your data.

**Comment by Peleg** [comment]
- my use case for having access to the hwm is a library which consumes the response's body but doesn't want to interfere with the enduser's ability to read it as well. my library just clones the response before reading it so it can be read again by the enduser. whether it is actually read or not, is not within my control.

- wouldnt setting a higher hwm essentially mean storing the stream in memory?

**Comment by bitinn** [comment]
One thing you can do is to wrap `res.body` (which is a Transform stream) with a stream of larger buffer. `text()`, `json()` will still happily consume `res.body` as long as they are a valid readable stream. End-user wouldn't have to know.

Setting a higher `highWaterMark` doesn't give you the actual control of the data itself. In most cases where you set `highWaterMark`, you want the full data anyway, so why not just buffer it.

TL;DR: your users will have to learn back pressure at some point and setting `highWaterMark` is just a workaround, never a solution. I am sorry but I have to put it bluntly.
